crl/updater/updater.go

```diff
@@ -141,64 +143,88 @@ func (cu *crlUpdater) Run(ctx context.Context) {
+			err := cu.Tick(ctx, atTime)
+			if err != nil {
+				// We only log, rather than return, so that the long-lived process can
+				// continue and try again at the next tick.
+				cu.log.AuditErrf("tick at time %s failed: %s", atTime, err)
+			}
 		case <-ctx.Done():
 			ticker.Stop()
-			return
+			return ctx.Err()
 		}
 	}
 }
 
-// TODO(#6261): Unify all error messages to identify the shard they're working
-// on as a JSON object including issuer, crl number, and shard number.
-
-func (cu *crlUpdater) Tick(ctx context.Context) {
-	atTime := cu.clk.Now()
-	result := "success"
+// Tick runs the entire update process once immediately. It processes each
+// configured issuer serially, and processes all of them even if an early one
+// encounters an error. All errors encountered are returned as a single combined
+// error at the end.
+func (cu *crlUpdater) Tick(ctx context.Context, atTime time.Time) (err error) {
 	defer func() {
+		// Each return statement in this function assigns its returned value to the
+		// named return variable `err`. This deferred function closes over that
+		// variable, and so can reference and modify it.
+		result := "success"
+		if err != nil {
+			result = "failed"
+			err = fmt.Errorf("tick %s: %w", atTime.Format(time.RFC3339Nano), err)
+		}
 		cu.tickHistogram.WithLabelValues("all", result).Observe(cu.clk.Since(atTime).Seconds())
 	}()
 	cu.log.Debugf("Ticking at time %s", atTime)
 
-	for id, iss := range cu.issuers {
+	var errStrs []string
+	for id := range cu.issuers {
 		// For now, process each issuer serially. This keeps the worker pool system
 		// simple, and processing all of the issuers in parallel likely wouldn't
 		// meaningfully speed up the overall process.
 		err := cu.tickIssuer(ctx, atTime, id)
 		if err != nil {
-			cu.log.AuditErrf(
-				"tick for issuer %s at time %s failed: %s",
-				iss.Subject.CommonName,
-				atTime.Format(time.RFC3339Nano),
-				err)
-			result = "failed"
+			errStrs = append(errStrs, err.Error())
 		}
 	}
+
+	if len(errStrs) != 0 {
+		return errors.New(strings.Join(errStrs, "; "))
+	}
+	return nil
 }
 
-// tickIssuer performs the full CRL issuance cycle for a single issuer cert.
-func (cu *crlUpdater) tickIssuer(ctx context.Context, atTime time.Time, issuerNameID issuance.IssuerNameID) error {
+// tickIssuer performs the full CRL issuance cycle for a single issuer cert. It
+// processes all of the shards of this issuer's CRL concurrently, and processes
+// all of them even if an early one encounters an error. All errors encountered
+// are returned as a single combined error at the end.
+func (cu *crlUpdater) tickIssuer(ctx context.Context, atTime time.Time, issuerNameID issuance.IssuerNameID) (err error) {
 	start := cu.clk.Now()
-	result := "success"
 	defer func() {
+		// Each return statement in this function assigns its returned value to the
+		// named return variable `err`. This deferred function closes over that
+		// variable, and so can reference and modify it.
```

Comment:
I don't think this trick works. Sure, you can modify `err`, but it has no impact on the return value. E.g. https://go.dev/play/p/XuowPVgTOv3

```go
package main

import (
	"errors"
	"fmt"
)

func foo() error {
	var err error
	defer func() {
		err = errors.New("sike")
	}()
	err = errors.New("I thought I was returning this error")
	return err
}

func main() {
	fmt.Println(foo())
}
```

Reply:
It does work, but _only_ because the return variable is _named_, which we usually avoid in Boulder code: https://go.dev/play/p/kIUb3KIksLZ
```go
package main

import (
	"errors"
	"fmt"
)

func foo() (err error) {
	defer func() {
		err = errors.New("sike")
	}()
	err = errors.New("I thought I was returning this error")
	return err
}

func main() {
	fmt.Println(foo())
}
```

Reply:
Ohhh wild, thanks for explaining.

Reply:
Yeah it's... not my favorite feature of go :P But it's certainly useful here. We can definitely improve the comments (I say "the named return variable `err`", but that's probably super easy to read as "the return value named `err`") to make this more explicit.